File: apps/web/lib/api/companionVoice.ts
import { apiGet, apiPost, queryString, type QueryParams } from "./client";
import type { CompanionVoiceSessionBundle, PaginatedItems } from "@/lib/types";

function normalizeVoiceSession(data: Partial<CompanionVoiceSessionBundle> | null): CompanionVoiceSessionBundle {
  const item = data ?? {};
  return {
    id: item.id ?? "",
    user_id: item.user_id ?? "",
    realtime_session_id: item.realtime_session_id ?? "",
    co_presence_session_id: item.co_presence_session_id ?? null,
    speaker_companion_id: item.speaker_companion_id ?? "",
    speaker_realtime_participant_id: item.speaker_realtime_participant_id ?? null,
    voice_profile_id: item.voice_profile_id ?? null,
    session_status: item.session_status ?? "created",
    transcript_retention_policy: item.transcript_retention_policy ?? "ephemeral",
    memory_write_policy: item.memory_write_policy ?? "candidate_review",
    allow_multi_speaker: item.allow_multi_speaker ?? false,
    permission_snapshot_json: item.permission_snapshot_json ?? {},
    voice_runtime_json: item.voice_runtime_json ?? {},
    turns: item.turns ?? [],
    stt_events: item.stt_events ?? [],
    tts_events: item.tts_events ?? [],
    turn_taking_events: item.turn_taking_events ?? [],
    interruptions: item.interruptions ?? [],
    persona_guard_runs: item.persona_guard_runs ?? [],
    created_at: item.created_at ?? null,
    updated_at: item.updated_at ?? null,
    metadata: item.metadata ?? {},
  };
}

export function listCompanionVoiceSessions(params?: QueryParams) {
  return apiGet<PaginatedItems<CompanionVoiceSessionBundle>>(
    `/companion-voice-sessions${queryString(params)}`,
  ).then((page) => ({ ...page, items: (page.items ?? []).map(normalizeVoiceSession) }));
}

export function createCompanionVoiceSession(data: Record<string, unknown>) {
  return apiPost<CompanionVoiceSessionBundle>("/companion-voice-sessions", data).then(normalizeVoiceSession);
}

export function getCompanionVoiceSession(voiceSessionId: string) {
  return apiGet<CompanionVoiceSessionBundle>(`/companion-voice-sessions/${voiceSessionId}`).then(normalizeVoiceSession);
}

export function recordSttPartial(voiceSessionId: string, data: Record<string, unknown>) {
  return apiPost(`/companion-voice-sessions/${voiceSessionId}/stt/partial`, data);
}

export function recordSttFinal(voiceSessionId: string, data: Record<string, unknown>) {
  return apiPost(`/companion-voice-sessions/${voiceSessionId}/stt/final`, data);
}

export function recordTtsEvent(voiceSessionId: string, data: Record<string, unknown>) {
  return apiPost(`/companion-voice-sessions/${voiceSessionId}/tts-events`, data);
}

export function decideTurnTaking(voiceSessionId: string, data: Record<string, unknown>) {
  return apiPost(`/companion-voice-sessions/${voiceSessionId}/turn-taking`, data);
}

export function recordVoiceInterruption(voiceSessionId: string, data: Record<string, unknown>) {
  return apiPost(`/companion-voice-sessions/${voiceSessionId}/interruptions`, data);
}

export function runVoicePersonaGuard(voiceSessionId: string, data: Record<string, unknown>) {
  return apiPost(`/companion-voice-sessions/${voiceSessionId}/persona-guard`, data);
}

File: services/agent-api/app/api/routes/companion_voice.py
"""Deferred companion voice service contracts."""

import uuid

from fastapi import APIRouter, Query

from app.schemas.common import err, ok, paginated_ok
from app.services import companion_voice_service

router = APIRouter(tags=["Companion Voice"])


def _get_seed_user_id() -> uuid.UUID:
    from app.db.models import User

    s = companion_voice_service.get_session()
    u = s.query(User).first()
    s.close()
    return u.id if u else uuid.uuid4()


@router.get("/companion-voice-sessions")
def list_companion_voice_sessions(
    user_id: str | None = Query(None),
    realtime_session_id: str | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    result = companion_voice_service.list_voice_sessions(
        user_id=uuid.UUID(user_id) if user_id else None,
        realtime_session_id=uuid.UUID(realtime_session_id) if realtime_session_id else None,
        page=page,
        page_size=page_size,
    )
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/companion-voice-sessions")
def create_companion_voice_session(body: dict):
    user_id = uuid.UUID(body["user_id"]) if body.get("user_id") else _get_seed_user_id()
    data = companion_voice_service.create_voice_session(user_id, body or {})
    if not data:
        return err("COMPANION_VOICE_CREATE_FAILED", "Unable to create companion voice session")
    return ok(data)


@router.get("/companion-voice-sessions/{voice_session_id}")
def get_companion_voice_session(voice_session_id: str):
    data = companion_voice_service.get_voice_session_bundle(uuid.UUID(voice_session_id))
    if not data:
        return err("COMPANION_VOICE_SESSION_NOT_FOUND", "Companion voice session not found")
    return ok(data)


@router.post("/companion-voice-sessions/{voice_session_id}/stt/partial")
def record_stt_partial(voice_session_id: str, body: dict):
    data = companion_voice_service.record_stt_partial(uuid.UUID(voice_session_id), body or {})
    if not data:
        return err("COMPANION_VOICE_SESSION_NOT_FOUND", "Companion voice session not found")
    return ok(data)


@router.post("/companion-voice-sessions/{voice_session_id}/stt/final")
def record_stt_final(voice_session_id: str, body: dict):
    data = companion_voice_service.record_stt_final(uuid.UUID(voice_session_id), body or {})
    if not data:
        return err("COMPANION_VOICE_SESSION_NOT_FOUND", "Companion voice session not found")
    return ok(data)


@router.post("/companion-voice-sessions/{voice_session_id}/tts-events")
def record_tts_event(voice_session_id: str, body: dict):
    data = companion_voice_service.record_tts_event(uuid.UUID(voice_session_id), body or {})
    if not data:
        return err("COMPANION_VOICE_SESSION_NOT_FOUND", "Companion voice session not found")
    return ok(data)


@router.post("/companion-voice-sessions/{voice_session_id}/turn-taking")
def decide_turn_taking_state(voice_session_id: str, body: dict):
    data = companion_voice_service.decide_turn_taking_state(uuid.UUID(voice_session_id), body or {})
    if not data:
        return err("COMPANION_VOICE_SESSION_NOT_FOUND", "Companion voice session not found")
    return ok(data)


@router.post("/companion-voice-sessions/{voice_session_id}/interruptions")
def record_interruption(voice_session_id: str, body: dict):
    data = companion_voice_service.record_interruption(uuid.UUID(voice_session_id), body or {})
    if not data:
        return err("COMPANION_VOICE_SESSION_NOT_FOUND", "Companion voice session not found")
    return ok(data)


@router.post("/companion-voice-sessions/{voice_session_id}/persona-guard")
def run_voice_persona_guard(voice_session_id: str, body: dict):
    data = companion_voice_service.run_voice_persona_guard(uuid.UUID(voice_session_id), body or {})
    if not data:
        return err("COMPANION_VOICE_SESSION_NOT_FOUND", "Companion voice session not found")
    return ok(data)
